port/3ds.h

```diff
@@ -0,0 +1,761 @@
+  assert(b == a);
+  glBlendEquation(GL_FUNC_ADD);
+
+  assert(c == GPU_SRC_ALPHA);
+  assert(d == GPU_ONE_MINUS_SRC_ALPHA);
+  assert(e == GPU_SRC_ALPHA);
+  assert(f == GPU_ONE_MINUS_SRC_ALPHA);
+  glBlendFuncSeparate(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA, GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+}
+static void GPU_SetAlphaTest(bool a, int b, uint8_t c) {
+  //FIXME: Implement
+  assert(a == false);
+  assert(b == GPU_ALWAYS);
+  assert(c == 0x00);
+}
+static void GPU_SetAttributeBuffers(int a, ...) {
+  //FIXME: Implement
+
+  //FIXME: GPU_SetAttributeBuffers(3, (u32*)osConvertVirtToPhys((u32)texData),
+//			GPU_ATTRIBFMT(0, 3, GPU_FLOAT)|GPU_ATTRIBFMT(1, 2, GPU_FLOAT)|GPU_ATTRIBFMT(2, 3, GPU_FLOAT),
+//			0xFFC, 0x210, 1, (u32[]){0x00000000}, (u64[]){0x210}, (u8[]){3});  
+
+#warning Missing GPU_SetAttributeBuffers
+  //FIXME_HACK()
+}
+static void GPU_SetDepthTestAndWriteMask(int a, int b, int c) {
+  //FIXME: Implement
+  assert(a == true);
+  glEnable(GL_DEPTH_TEST);
+
+  assert(b == GPU_GREATER);
+  glDepthFunc(GL_GREATER);
+
+  assert(c == GPU_WRITE_ALL); //FIXME: glDepthMask?
+}
+static void GPU_SetTexEnv(int a, int b, int c, int d, int e, int f, int g, u32 h) {
+/*
+GPU_SetTexEnv(0, 
+84			GPU_TEVSOURCES(GPU_TEXTURE0, GPU_PRIMARY_COLOR, GPU_PRIMARY_COLOR), 
+85			GPU_TEVSOURCES(GPU_TEXTURE0, GPU_PRIMARY_COLOR, GPU_PRIMARY_COLOR),
+86			GPU_TEVOPERANDS(0,0,0), 
+87			GPU_TEVOPERANDS(0,0,0), 
+(gdb) list
+88			GPU_MODULATE, GPU_MODULATE, 
+89			0xFFFFFFFF);
+*/
+  //FIXME: Implement!
+}
+static void GPU_SetTextureEnable(int a) {
+  assert(a == GPU_TEXUNIT0);
+  glEnable(GL_TEXTURE_2D);
+}
+static void GPU_SetTexture(int a, u32* b, int width, int height, int filter, int format) {
+  assert(a == GPU_TEXUNIT0);
+//  assert(filter == (GPU_TEXTURE_MAG_FILTER(GPU_NEAREST)|GPU_TEXTURE_MIN_FILTER(GPU_NEAREST)));
+//FIXME: filter also GPU_TEXTURE_MAG_FILTER(GPU_NEAREST)|GPU_TEXTURE_MIN_FILTER(GPU_NEAREST)|GPU_TEXTURE_WRAP_S(GPU_REPEAT)|GPU_TEXTURE_WRAP_T(GPU_REPEAT)
+  assert(format == GPU_RGBA8);
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, b);
+}
+static void GPU_SetUniform(u32 reg, u32* values, int count) {//FIXME: Implement
+  //printf("Setting uniform %d to 0x%08X [%d values]\n", reg, values[0], count);
+
+  float* f = (float*)values;
+
+  // Hack via u32 and SHDR_GetUniformRegister
+  char* name = (char*)reg;
+
+  if (!strcmp(name, "projection")) {
+    glMatrixMode(GL_PROJECTION);
+    assert(count == 4);
+    float m[4*4];
+    for(int i = 0; i < count; i++) {
+      for(int j = 0; j < 4; j++) {
+        m[i * 4 + j] = f[i * 4 + (3 - j)];
+      }
+    }
+    glLoadIdentity();
+    glMultTransposeMatrixf(m);
+
+#if 1    
+    glLoadIdentity();
+    gluPerspective(80.0f, 240.0f/400.0f, 0.01f, 1000.0f); //45.0f, 480.0f / 400.0f, -100.0f, 100.0f);
+#endif
```

Comment:
Super ugly hack. Should compare resulting matrices - the depth is wrong with smeas code.